src/nbl/ext/FFT/FFT.cpp

```diff
@@ -0,0 +1,368 @@
+{
+	assert(core::isPoT(maxPaddedDimensionSize));
+
+	const char* sourceFmt =
+R"===(#version 430 core
+
+#define USE_SSBO_FOR_INPUT %u
+
+// WorkGroup Size
+
+#ifndef _NBL_GLSL_EXT_FFT_BLOCK_SIZE_X_DEFINED_
+#define _NBL_GLSL_EXT_FFT_BLOCK_SIZE_X_DEFINED_ %u
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_BLOCK_SIZE_Y_DEFINED_
+#define _NBL_GLSL_EXT_FFT_BLOCK_SIZE_Y_DEFINED_ 1
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_BLOCK_SIZE_Z_DEFINED_
+#define _NBL_GLSL_EXT_FFT_BLOCK_SIZE_Z_DEFINED_ 1
+#endif
+
+#define _NBL_GLSL_WORKGROUP_SIZE_ (_NBL_GLSL_EXT_FFT_BLOCK_SIZE_X_DEFINED_*_NBL_GLSL_EXT_FFT_BLOCK_SIZE_Y_DEFINED_*_NBL_GLSL_EXT_FFT_BLOCK_SIZE_Z_DEFINED_)
+
+layout(local_size_x=_NBL_GLSL_EXT_FFT_BLOCK_SIZE_X_DEFINED_, local_size_y=_NBL_GLSL_EXT_FFT_BLOCK_SIZE_Y_DEFINED_, local_size_z=_NBL_GLSL_EXT_FFT_BLOCK_SIZE_Z_DEFINED_) in;
+ 
+#define _NBL_GLSL_EXT_FFT_MAX_DIM_SIZE_ %u
+#define _NBL_GLSL_EXT_FFT_MAX_ITEMS_PER_THREAD %u
+ 
+#define _NBL_GLSL_EXT_FFT_GET_DATA_DEFINED_
+#define _NBL_GLSL_EXT_FFT_SET_DATA_DEFINED_
+#include "nbl/builtin/glsl/ext/FFT/fft.glsl"
+
+// Input Descriptor
+
+struct nbl_glsl_ext_FFT_input_t
+{
+	vec2 complex_value;
+};
+
+#ifndef _NBL_GLSL_EXT_FFT_INPUT_SET_DEFINED_
+#define _NBL_GLSL_EXT_FFT_INPUT_SET_DEFINED_ 0
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_INPUT_BINDING_DEFINED_
+#define _NBL_GLSL_EXT_FFT_INPUT_BINDING_DEFINED_ 0
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_INPUT_DESCRIPTOR_DEFINED_
+
+#if USE_SSBO_FOR_INPUT > 0
+#define _NBL_GLSL_EXT_FFT_INPUT_DESCRIPTOR_DEFINED_
+layout(set=_NBL_GLSL_EXT_FFT_INPUT_SET_DEFINED_, binding=_NBL_GLSL_EXT_FFT_INPUT_BINDING_DEFINED_) readonly restrict buffer InputBuffer
+{
+	nbl_glsl_ext_FFT_input_t inData[];
+};
+#else 
+layout(set=_NBL_GLSL_EXT_FFT_INPUT_SET_DEFINED_, binding=_NBL_GLSL_EXT_FFT_INPUT_BINDING_DEFINED_) uniform sampler2D inputImage;
+#endif
+
+#endif
+
+// Output Descriptor
+
+struct nbl_glsl_ext_FFT_output_t
+{
+	vec2 complex_value;
+};
+
+
+#ifndef _NBL_GLSL_EXT_FFT_OUTPUT_SET_DEFINED_
+#define _NBL_GLSL_EXT_FFT_OUTPUT_SET_DEFINED_ 0
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_OUTPUT_BINDING_DEFINED_
+#define _NBL_GLSL_EXT_FFT_OUTPUT_BINDING_DEFINED_ 1
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_OUTPUT_DESCRIPTOR_DEFINED_
+#define _NBL_GLSL_EXT_FFT_OUTPUT_DESCRIPTOR_DEFINED_
+layout(set=_NBL_GLSL_EXT_FFT_OUTPUT_SET_DEFINED_, binding=_NBL_GLSL_EXT_FFT_OUTPUT_BINDING_DEFINED_) restrict buffer OutputBuffer
+{
+	nbl_glsl_ext_FFT_output_t outData[];
+};
+#endif
+
+// Get/Set Data Function
+
+vec2 nbl_glsl_ext_FFT_getData(in uvec3 coordinate, in uint channel)
+{
+	vec2 retValue = vec2(0, 0);
+#if USE_SSBO_FOR_INPUT > 0
+	uvec3 dimension = pc.dimension;
+	uint index = channel * (dimension.x * dimension.y * dimension.z) + coordinate.z * (dimension.x * dimension.y) + coordinate.y * (dimension.x) + coordinate.x;
+	retValue = inData[index].complex_value;
+#else
+	vec4 texelValue= texelFetch(inputImage, ivec2(coordinate.xy), 0);
```

Comment:
does `texelFetch` respect sampler modes (when the unnormalized coords are outside the texture range)?

Please find the answer as to what happens, in the GLSL spec and paste here.

Reply:
https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/texelFetch.xhtml

It isn't mentioned in the OpenGL Reference explicitly. 
but the input is "sampler" which I believe Vulkan calls "CombinedImageSampler".
I may be wrong here but then it implicitly should honor the properties of the sampler "combined" with the image.

Reply:
Let m find that for you

> The integer coordinates (i; j; k) passed to texelFetch are used to pointsample
the texture image. The level of detail accessed is computed by adding the
specified level-of-detail parameter lod to the base level of the texture, levelbase.
Texel fetch proceeds similarly to the steps described for texture access in section
11.1.3.5, with the exception that none of the operations controlled by sampler
object state are performed, including:
 level of detail clamping;
 texture wrap mode application;
 filtering (however, a mipmapped minification filter is required to access any
level of detail other than the base level);
 depth comparison.
The steps that are performed are:
 validation of texel coordinates as described below, including the computed
level-of-detail, (i; j; k), the specified level for array textures, and texture
completeness;
 sRGB conversion of fetched values as described in section 8.24;
 conversion to base color Cb;
 component swizzling.
The results of texelFetch builtins are undefined if any of the following
conditions hold:
 the computed level of detail is less than the texture’s base level (levelbase) or
greater than the maximum defined level, q (see section 8.14.3)
 the computed level of detail is not the texture’s base level and the texture’s
minification filter is NEAREST or LINEAR
 the layer specified for array textures is negative, or greater than or equal to
the number of layers in the array texture
the texel coordinates (i; j; k) refer to a texel outside the defined extents of
the computed level of detail, where any of
i < 0 i  ws
j < 0 j  hs
k < 0 k  ds
and ws, hs, and ds refer to the width, height, and depth of the image, as
defined in section 8.5.3.

Reply:
You cannot use `texelFetch` when you expect wrapping.

Use `textureLod` everywhere instead